warl_proposal/csr_template_draft1.yaml

```diff
@@ -0,0 +1,130 @@
+############################################ Template description ############################################
+# csr template for csrs with subfields
+<name>:                                   # name of the csr
+  description: <text>                     # textual description of the csr
+  address: <hex>                          # address of the CSR
+  priv_mode: <D/M/H/S/U>                  # privilege mode that owns the register
+  reset-val: <hex>                        # Reset value of the register. This an accumulation of the all reset values of the 
+                                          # sub-field 
+  rv32:                                   # this node and its subsequent fields can exist if [M/S/U]XL value can be 1
+    exists: <boolean>                     # indicates if the csr exists in rv32 mode or not. When False, all fields below 
+                                          # will be trimmed off in the checked yaml. False also indicates that 
+                                          # access-exception should be generated
+    subfields:                            # a quick summary of the list of all subfields of the csr including a list of 
+                                          # WPRI fields of the csr.
+      - <subfield_name1>
+      - <subfield_name2>
+      - wpri                              # contains a list of all WPRI fields within the csr. Can be empty as well
```

Comment:
Should WPRI fields be strictly defined in the privileged spec? or should any undefined bit position be implied as WPRI?

Reply:
WPRI signifies that the spec has reserved these fields for future use and an implementation is not allowed to use them for customization. Later versions of the spec can start using these fields for define more functionality, (like how the TEE guys are modifying the pmp defs). So the WPRI fields should be captured as they are mandated by the spec.

The spec also defines certains fields left for platform use (like the bits >16 in mip and mie). The implementation can choose to define these or not. But I feel the WPRI fields should be captured in the spec.

Reply:
I agree with what you said. If it's mandated in the spec, the checker should check if user used any WPRI field for other purpose. I just feel that do we need user to input this information again here?

Reply:
I think this one should be resolved as well. As mentioned in the other comment, this is the input to the checker, not necessary the information user needs to enter.

Reply:
As i mentioned here: https://github.com/riscv/riscv-config/pull/5#discussion_r313718318
( apologize for not making this clear before).. 

The user does not have to fill all the fields up. The checker will fill most of these automatically. What you are seeing here is the final checker generated yaml which can be used for multiple use-cases. 

I think I should create an example to show this. For the mtvec example given in the file, the user needs to provide a yaml with only the following:

```
mtvec:                                                                                              
                                                                        
  reset-val: 0x00000001                                                                             
  rv32:                                                                                             
    exists: True                                                                                                    
    mode:                                                                                           
      implemented: True                                                                             
      reset-val: 0x1                                                                                
      type:                                                                                         
        WARL:                                                                                       
          dependency_variables: []                                                                  
          behavior:                                                                                 
            - func1:                                                                                
                dependency_values: []                                                               
                warl_func:                                                                          
                  mode: "Unchanged"                                                                 
                  legal: [[0x3,0x0,0x0,0x1]]                                                        
    base:                                                                                           
      implemented: True                                                                             
      reset-val: 0x0                                                                                
      type:                                                                                         
        WARL:                                                                                       
          dependency_variables: [mtvec::mode]                                                       
          behavior:                                                                                 
            - func1:                                                                                
                dependency_values: [0]                                                              
                warl_func:                                                                          
                  mode: "Unchanged"                                                                 
                  legal: [[0x20000000], [0x20004000]]             # can take only 2 fixed values in direct mode.
            - func2:                                                                                
                dependency_values: [1]                                                              
                warl_func:                                                                          
                  mode: "Unchanged"                                                                 
                  legal: [[0xFFFFFFC0, 0x0, 0x00000000,0x20000000]] # 256 byte aligned values only in vectored mode.
  rv64:                                                                                             
    exists: False                                                                                   
  rv128:                                                                                            
    exists: False                                          
```

The checker would parse this, perform specific checks on each field and generate the output you see in csr_template_draft1.yaml.. does that make sense?

Reply:
yes, thanks for clarification.